docs/enhancements/clusterpool-inventory.md

```diff
@@ -0,0 +1,175 @@
+      - name: baz-install-config-patch
+```
+
+and InstallConfigPatch CR will look like
+```yaml
+apiVersion: v1
+kind: InstallConfigPatch
+metadata:
+  name: foo-install-config-patch
+  namespace: my-project
+spec:
+  patch: <base64 encoded json patch>    
+status:
+  clusterDeploymentRef: 
+    name: foo
+  conditions:
+    - lastProbeTime: "2020-11-05T14:49:26Z"
+      lastTransitionTime: "2020-11-05T14:49:26Z"
+      message: patch is currently in use by cluster deployment foo of clusterpool foo-pool
+      reason: PatchInUse
+      status: "false"
+      type: PatchAvailable
+```
+
+if DNS is configured with the name `foo`, The base 64 encoded json content to patch vSphere install config will be as follows
+
+```json
+[
+  { "op": "replace", "path": "metadata/name", "value": "foo" }
+]
+```
+
+When adding a ClusterDeployment, if such an `Inventory` is present, ClusterPool controller will:
+- Load up the inventory list.
+- For each reference it will do a GET to fetch InstallConfigPatch and check the if the PatchAvailable condition is true.
+- For the first InstallConfigPatch that is available, it will use the json patch in the `spec.patch` field to update the default install config generated by clusterpool. It will also update the status with reference to the cluster deployment that is using it and update the `PatchAvailable` condition to false.
+- Set the `spec.clusterPoolReference.installConfigPatchRef`field in the ClusterDeployment with a reference to InstallConfigPatch CR.
+
+Absent an `Inventory`, ClusterPool will continue to use the generated default install config as it does today.
+
+When ClusterDeployment is deleted, ClusterPool controller will: 
+- Find out if ClusterDeployment used an InstallConfigPatch from `spec.ClusterPoolReference.InstallConfigPatchRef`field.
+- If yes, we simply update InstallConfigPatch, setting the `spec.clusterDeploymentRef` to nil and `PatchAvailable` to true.  
+
+### How To Use
+For the VSphere case, this allows the administrator to:
+- Preconfigure DNS with following entries (assuming cluster name is `foo`)
+    ```
+    10.0.0.10  api.foo.example.com
+    10.0.0.11  apps.foo.example.com
+    10.0.0.12  console.foo.example.com
+    ```
+- Create a InstallConfigPatch CR to patch `spec.metadata.name` field of the default install config generated by clusterpool controller. Please refer the section above of a sample CR. The base64 encoded json content in `spec.patch` field should be as follows
+    ```json
+    [
+      { "op": "replace", "path": "metadata/name", "value": "foo" }
+    ]
+    ```
+- Add the name of InstallConfigPatch CR to`clusterPool.spec.inventory.installConfigPatches` list. For InstallConfigPatch with a name `foo-install-config-patch` the clusterpool should be configured as follows
+    ```yaml
+    spec:
+      inventory:
+        installConfigPatches:
+          - name: foo-install-config-patch
+    ```
+  
+### Validation
+Webhook validation will ensure that for a clusterpool
```

Comment:
It feels like it should just be a runtime condition if a user sets maxSize and size to nonsensical values. If you put size 5, and inventory contains 2 items, then you get 2 with a condition saying "cant get to size 5".

Reply:
agreed, I will make that change. Also, maxSize (if set) should be equal to inventory size.